design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,206 @@
+
+### Memory Hot Plug
+
+The logic of memory hot plug will be similar to CPU hot plug, with the exception of the fields names.
+`spec.template.spec.domain.memory.maxGuest` will become mutable and will remain optional.
+A default value for the `MaxGuest` memory value will continue to be internally calculated based on the cluster setting of `maxHotplugRatio` 
+
+As part of this change, the `MaxGuest` field will be renamed to `MaxMemory`.
+
+All other hot plug behavior wil remain the same.
+
+## Drawbacks / Limitations
+
+-   No cluster- and VM-level fine-grained control over rollout settings of specific features.
+-   No VM-specific controll over the rollout strategy.
+
+## API 
+
+ 
+### Before the changes in this proposal:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-cirros
+    name: vm-cirros
+spec:
+  liveUpdateFeatures:
+    cpu:
+      maxSockets: 8
+    memory:
+      maxGuest: 256Mi
+  running: false
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      domain:
+        cpu:
+          sockets: 2
+          cores: 4
+          threads: 2
+        memory:
+          guest: 128Mi
+```
+
+### After the changes in this proposal:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-cirros
+name: vm-cirros
+spec:
+  running: false
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      domain:
+        cpu:
+          sockets: 2
+          cores: 4
+          threads: 2
+          maxSockets: 8  # optional
+        memory:
+          guest: 128Mi
+          maxGuest: 256Mi  # optional
+```
+### CPU Hot plug example
+ Before CPU hot plug:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  template:
+    spec:
+      domain:
+        cpu:
+          sockets: 2
+          cores: 1
+          threads: 1
+```
+
+After CPU hot plug:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  template:
+    spec:
+      domain:
+        cpu:
+          sockets: 3  # This value changed from '2' to '3', this is the only change
+          cores: 1
+          threads: 1
+```
+
+### Status API
+
+A condition called `RestartRequired` will be added to the VM requesting a restart if the changes cannot be immediately applied.
+The condition will contain the reason for not being able to reach the desired state requested in the VM spec.
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  template:
+    spec:
+      domain:
+        cpu:
+          sockets: 42  # Changed to an unreasonable value, which does require a restart
+          cores: 1
+          threads: 1
+status:
+  conditions:
+    - type: RestartRequired
+      value: true
+      reason: 'The number of sockets is too high and requires a VM restart in order to become effective".
+```
+
+## Update/Rollback Compatibility
```

Comment:
note: for the future, it would be great to have a testing section. Approvers could ack/nack functional tests, unit tests, and such. This could help us correctly test things, note this is currently a problem recognized by many approvers.

Reply:
@xpivarc can you follow up with a PR to add this to the design template?